#[187364724] - A successfully registered user should receive an email to verify their account email

AngeloChristian1 opened the pull request:
**What does this PR do**
Users need to verify their email to avoid spams

**Description of Task to be completed?**
- Added nodemailer to help in sending emails
- Created a reusable component for sending emails
- Updated signup and login controllers to handle email verification on signup and check for verification status on login
- Added email verifying component to handle email verification logic
- Added endpoints for email verification `/user/verify/:token`

 **How should this be manually tested?**
- Fire up your application or use the hosted server
- You can use use swagger docs to test it by going to `/users/signup`
- Add your real email address and all necessary information for signup
- If successful, check your gmail or any email platform you use (for the email you used during signup)
- Click on the verification link provided
- Test the verification on swagger on `/users/verify/:token`

![Screenshot 2024-04-30 113339](https://github.com/atlp-rwanda/e-commerce-bitcrafters-bn/assets/39400022/cf32f96f-dfef-4581-88ad-bfc061949b52)
![Screenshot 2024-04-30 113406](https://github.com/atlp-rwanda/e-commerce-bitcrafters-bn/assets/39400022/25d21201-e8c1-4637-ba9c-27d63912b286)
![Screenshot 2024-04-30 113452](https://github.com/atlp-rwanda/e-commerce-bitcrafters-bn/assets/39400022/8d5b24f6-d2cf-4661-99a7-0bf824d767c4)
![Screenshot 2024-04-30 113720](https://github.com/atlp-rwanda/e-commerce-bitcrafters-bn/assets/39400022/3da72b3c-be97-4c78-9893-c66197433bfb)
![Screenshot 2024-04-30 113746](https://github.com/atlp-rwanda/e-commerce-bitcrafters-bn/assets/39400022/c49e1cea-56da-4e16-9f33-e936c2226d05)







 **What are the relevant pivotal tracker/Trello stories?**
#187364724

EricNiyo73 reviewed (changes requested):
I think this VScode folder , it is not necessary. So delete it.
![image](https://github.com/atlp-rwanda/e-commerce-bitcrafters-bn/assets/106761137/bcd4aa11-0e8b-4dcd-8b81-7fbc4e42b639)

didasmbalanya commented on src/controllers/UserController.ts:
```diff
@@ -32,13 +36,28 @@ export default class UserController {
 
     await newUser.save()
     const token = jwt.sign({ id: newUser.id }, process.env.JWT_SECRET)
+    const verificationToken = generateToken({
+      email: newUser.email,
+      username: newUser.username,
+    })
+    const baseUrl = `${process.env.NODEMAILER_BASE_URL}users/verify/${token}`
+
+    const html = verificationsEmailTemplate(newUser, baseUrl)
+
+    await sendMail(email, 'Verify Account', html)
+    const newVerificationToken = await Token.create({
+      userId: newUser.dataValues.id,
+      token: verificationToken,
+    })
```
What is the difference between verificationnToken and newVerification token

didasmbalanya commented on src/controllers/UserController.ts:
```diff
@@ -32,13 +36,28 @@ export default class UserController {
     })
 
     await newUser.save()
     const token = jwt.sign({ id: newUser.id }, process.env.JWT_SECRET)
```
Please move this to utilities or somewhere it can be reused

didasmbalanya commented on src/controllers/UserController.ts:
```diff
@@ -32,13 +36,28 @@ export default class UserController {
+      token: verificationToken,
+    })
     return res.status(201).json({
       jwt: token,
```
A signup user should not be automatically signed in.

didasmbalanya commented on src/controllers/authentication.ts:
```diff
@@ -0,0 +1,30 @@
+
+const verifyEmail = async (req: express.Request, res: express.Response) => {
+  try {
+    const { token } = req.params
```
Handle a case where the token param does not exist?

munezerobagira replied:
you can also use the validation schema to test this and  make user that use provide a invalid param

didasmbalanya commented on src/controllers/authentication.ts:
```diff
@@ -0,0 +1,30 @@
+      },
+    })
+    if (!user) {
+      return res.status(404).send({ message: 'Invalid Link' })
```
Bad request, Try again

didasmbalanya commented on src/controllers/authentication.ts:
```diff
@@ -0,0 +1,30 @@
+    if (!user) {
+      return res.status(404).send({ message: 'Invalid Link' })
+    }
+    await user.update({ verified: true })
+    await Token.destroy({ where: { id: myToken.dataValues.id } })
```
Use 
```suggestion
await Promise.all([
    await user.update({ verified: true }),
    await Token.destroy({ where: { id: myToken.dataValues.id } })
    ])
```

didasmbalanya commented on src/database/migrations/20240427140502-create-user.js:
```diff
@@ -1,51 +0,0 @@
-     */
-    await queryInterface.dropTable('Users')
-  },
-}
```
Do not edit, or delete a merged migration file

didasmbalanya commented on src/database/migrations/20240429100117-create-user.js:
```diff
@@ -0,0 +1,42 @@
+      userRole: {
+        type: Sequelize.STRING,
+      },
+      verified: {
+        allowNull: false,
+        type: Sequelize.BOOLEAN,
+        defaultValue: false,
+      },
```
revert to the old migration files and add a new one that adds a new column

munezerobagira replied:
you don't have to new table, instead you need to alter table of users, as migration of users alread exists
https://sequelize.org/docs/v6/other-topics/query-interface/
https://sequelize.org/docs/v6/other-topics/migrations/

AngeloChristian1 replied:
I created a new migration file for adding a new column, actually the migration file that I had had no effect as the table had already been created when we started using the same local database

didasmbalanya commented on src/helpers/hashPasssword.ts:
```diff
@@ -0,0 +1,18 @@
+
+dotenv.config()
+
+export const hashPassword = async (password: string) => {
+  const slatRounds = await bcrypt.genSalt(10)
+  const hashedPassword = await bcrypt.hash(password, slatRounds)
+  return hashedPassword
+}
+
+export const comparePassword = async (
+  password: string,
+  hashedPassword: string,
+) => {
+  const compare = await bcrypt.compare(password, hashedPassword)
+  return compare
+}
```
Nice

didasmbalanya commented on src/helpers/jwt_config.ts:
```diff
@@ -0,0 +1,11 @@
+/* eslint-disable @typescript-eslint/no-namespace */
+import jwt from 'jsonwebtoken'
+
+const generateToken = (data: object) => {
+  const token = jwt.sign(data, process.env.JWT_SECRET || 'USER-AUTH', {
+    expiresIn: process.env.JWT_EXPIRE_TIME || 30000,
+  })
+  return token
+}
+
+export default generateToken
```
Use one file naming system

didasmbalanya commented on src/helpers/sendEmail.ts:
```diff
@@ -0,0 +1,27 @@
+      html: mail,
+    })
+  } catch (error) {
+    //
```
Please handle this incomplete code

didasmbalanya commented on src/routes/userRoute.ts:
```diff
@@ -1,8 +1,10 @@
+import verifyEmail from '../controllers/authentication'
 
 const router = Router()
 router.post('/signup', UserController.signup)
+router.get('/verify/:token', verifyEmail)
```
Please add validation as a separate route handler/middleware

didasmbalanya reviewed (changes requested):
Looks good but there are a few things that you can improve

munezerobagira commented on src/controllers/LoginController.ts:
```diff
@@ -45,7 +34,10 @@ export default class LoginController {
     }
 
-    const token = generateToken(existingUser.id)
+    const token = generateToken({
```
you need to add the check to make sure that user who is not verified can not get a  login token

munezerobagira commented on src/controllers/UserController.ts:
```diff
@@ -1,8 +1,11 @@
-import jwt from 'jsonwebtoken'
 import User from '../database/models/userModel'
-import userSchema from '../validations/userValidation'
+// import userSchema from '../validations/userValidation'
```
```suggestion

```

didasmbalanya replied:
please remove comments

munezerobagira reviewed (commented):
Nice work but there are few things which we can improve

didasmbalanya commented on src/controllers/UserController.ts:
```diff
@@ -32,13 +31,27 @@ export default class UserController {
+
+    const html = verificationsEmailTemplate(newUser, baseUrl)
+
+    await sendMail(email, 'Verify Account', html)
+    const verificationToken = await Token.create({
+      userId: newUser.dataValues.id,
+      token,
+    })
```
use Promise.all() to run both concurrently

didasmbalanya commented on src/controllers/UserController.ts:
```diff
@@ -32,13 +31,27 @@ export default class UserController {
-      jwt: token,
-      message: 'Account Created successfully',
+      message: 'Account Created successfully, verify your email to countinue',
+      verificationToken,
```
What is this verification token for if its sent to the email as well ?

didasmbalanya commented on src/controllers/authentication.ts:
```diff
@@ -0,0 +1,37 @@
+
+const verifyEmail = async (req: express.Request, res: express.Response) => {
+  try {
+    const { token } = req.params
+
+    if (!token) {
+      return res.status(400).send({ message: 'Bad request' })
+    }
```
This can be done via a middleware since you'll have many API endpoints that will need Authentication

didasmbalanya commented on src/controllers/authentication.ts:
```diff
@@ -0,0 +1,37 @@
+    if (!token) {
+      return res.status(400).send({ message: 'Bad request' })
+    }
+    const myToken = await Token.findOne({ where: { token } })
```
So the reason we use a JWT is also to be able to verify the authenticity of the request, meaning we need to verify that the token is from the right use. jwt.verify()

didasmbalanya reviewed (changes requested):
Good job on the fixes. here a few more things to work on

munezerobagira commented on src/assets/emailTemplates/index.ts:
```diff
@@ -0,0 +1,34 @@
+/* eslint-disable import/prefer-default-export */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
```suggestion
```

munezerobagira commented on src/controllers/authentication.ts:
```diff
@@ -0,0 +1,57 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable require-jsdoc */
```
```suggestion

```

munezerobagira commented on src/middlewares/paramValidation.ts:
```diff
@@ -0,0 +1,24 @@
+/* eslint-disable import/prefer-default-export */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+/* eslint-disable require-jsdoc */
```
```suggestion

```

EricNiyo73 commented on src/routes/loginRoute.ts:
```diff
@@ -1,8 +1,8 @@
-
 import LoginController from '../controllers/LoginController'
+import { loginValidator } from '../middlewares/routeValidation'
 
 const router = Router()
-router.post('/login', LoginController.login)
+router.post('/login', loginValidator, LoginController.login)
 
```
Try to put this login into user Route. we don't need to have many route which is  common. i.e. If you are working with a user, we have use Route which has all endpoints of user like signup, login, get user ,delete User, verifyEmail etc . Which is the same as for products and so on.

EricNiyo73 commented on src/controllers/authentication.ts:
```diff
@@ -0,0 +1,53 @@
+import User from '../database/models/userModel'
+import Token from '../database/models/tokenModel'
+
+export const validateAndRetrieveToken = async (token: string) => {
+  try {
+    const decodedToken = jwt.verify(token, process.env.JWT_SECRET, {
+      ignoreExpiration: false,
+    }) as JwtPayload
+    const myToken = await Token.findOne({ where: { token } })
```
I think this function can be used as helpers not a controller, try to move it into helpers folder.

AngeloChristian1 replied:
The reason it is here is because it is specifically for this verify token I separated them because of the codeclimate issue of cognitive comprexity and for easy testing

EricNiyo73 commented on src/controllers/authentication.ts:
```diff
@@ -0,0 +1,53 @@
+  }
+}
+
+export const verifyEmail = async (
+  req: express.Request,
+  res: express.Response,
+) => {
+  try {
+    const { token } = req.params
```
let us have this function into user controller, and try to call validateAndRetrieveToken(token) to verify the token.

EricNiyo73 commented on src/helpers/sendEmail.ts:
```diff
@@ -0,0 +1,28 @@
+import nodemailer from 'nodemailer'
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const Mailer: any = async (email: string, subject: string, mail: string) => {
+  try {
+    const transporter = nodemailer.createTransport({
+      service: process.env.NODEMAILER_SERVICE,
+      host: process.env.NODEMAILER_HOST,
+      port: parseInt(process.env.NODEMAILER_EMAIL_PORT || '', 10),
+      secure: process.env.SECURE === 'true',
+      auth: {
+        user: process.env.NODEMAILER_USER,
+        pass: process.env.NODEMAILER_PASS,
+      },
+    })
+
+    transporter.sendMail({
+      from: process.env.NODEMAILER_USER,
+      to: email,
+      subject,
```
Nice configurations.

EricNiyo73 commented on src/controllers/LoginController.ts:
```diff
@@ -16,31 +18,63 @@ export default class LoginController {
+        const html = verificationsEmailTemplate(existingUser.username, baseUrl)
+
+        await Promise.all([
+          sendMail(email, 'Verify Your Account', html),
+          Token.create({
+            userId: existingUser.id,
+            token: newVerificationToken,
+          }),
+        ])
```
Yes it is good idea to store the token into the database, but suppose that, we have many like 4M users tried to register to our e-commerce app, and then they have not tried to verify the account, This means the token will still stored into our DB.
So you can add the way you can manage how the time you can store that token. or you can use Redis to store the token because it has the way you can schedule the time to be expired. 

the Token will automatically expire after the specified duration. Once the token expires, Redis will automatically remove it from the database.

EricNiyo73 commented on src/controllers/LoginController.ts:
```diff
@@ -16,31 +18,63 @@ export default class LoginController {
+        if (pendingVerification) {
+          return res.status(200).send({ message: 'Please verify your email' })
+        }
+
+        // Generate and send new verification email
+        const newVerificationToken = generateToken({
+          id: existingUser.id,
+          email: existingUser.email,
+        })
+        const baseUrl = `${process.env.NODEMAILER_BASE_URL}users/verify/${newVerificationToken}`
+        const html = verificationsEmailTemplate(existingUser.username, baseUrl)
+
+        await Promise.all([
+          sendMail(email, 'Verify Your Account', html),
```
By the way , this is good idea to resend a verification link to the user tried to login while they have not verified their account. Good job.

EricNiyo73 reviewed (changes requested):
use Redis to store your token and manage the time it can be expired and then be removed from DB.

EricNiyo73 commented on src/controllers/LoginController.ts:
```diff
@@ -18,62 +20,96 @@ export default class LoginController {
-          await sendMail(email, 'OTP VERIFICATION CODE', html)
+      }
+
+      if (existingUser.userRole === 'seller') {
+        const otp = `${Math.floor(1000 + Math.random() * 9000)}`
+        const html = otpEmailTemplate(existingUser.username, otp)
+        const otptoken = generateToken({
+          id: existingUser.id,
+          email: existingUser.email,
+          otp,
         })
+        await redisClient
+          .setEx(existingUser.email, 300, `${otp}=${otptoken}`)
+          .then(async () => {
+            await sendMail(email, 'OTP VERIFICATION CODE', html)
+          })
 
-      tokenPayload.otp = otp
-      const sellerTokenMessage =
-        'Email sent to your email. Please check your inbox messages and enter the OTP for verification'
-      return res.status(200).json({ message: sellerTokenMessage })
-    }
+        tokenPayload.otp = otp
+        const sellerTokenMessage =
+          'Email sent to your email. Please check your inbox messages and enter the OTP for verification'
+        return res.status(200).json({ message: sellerTokenMessage })
+      }
 
-    const token = generateToken(tokenPayload)
-    return res
-      .status(200)
-      .header('authorization', token)
-      .json({ jwt: token, message: 'Login successful' })
+      // Generate and return JWT token for authenticated user
+      const authToken = generateToken({
+        id: existingUser.id,
+        email: existingUser.email,
+        role: existingUser.userRole,
+      })
+      return res.status(200).send({ message: 'Login successful', authToken })
```
I think this token generation is the same, why have you changed this??

AngeloChristian1 committed:
Feature(Verfication Email): User will receive an email to verify their account

    1. Setting up email service with Nodemailer
    2. Created a reusable component for sending emails
    [Delivers #187364724]

munezerobagira commented on src/controllers/LoginController.ts:
```diff
@@ -18,62 +22,97 @@ export default class LoginController {
+          .send({ message: 'A verification email has been sent' })
+      }
+
+      const tokenPayload: {
+        id: number
+        userRole: string
+        email: string
+        otp?: string
+      } = {
         id: existingUser.id,
+        userRole: existingUser.userRole,
         email: existingUser.email,
```
```suggestion
      const tokenPayload = {
        id: existingUser.id,
        userRole: existingUser.userRole,
        email: existingUser.email,
```
I guess the type  will be infered automatically

munezerobagira commented on src/controllers/LoginController.ts:
```diff
@@ -18,62 +22,97 @@ export default class LoginController {
-          await sendMail(email, 'OTP VERIFICATION CODE', html)
+      }
+
+      if (existingUser.userRole === 'seller') {
```
```suggestion
      if (existingUser.userRole ===UserRole.Seller) {
```
We can use enum

munezerobagira reviewed (changes requested):
Nice work, but we can improve few things

munezerobagira commented:
Can you  update description on how test this  feature using deployed link( like one on the PR preview)
